### ISSUE TITLE
Login Starts Failing, Requiring Server Restart

### ISSUE THREAD

**Issue by msymons** [DependencyTrack/dependency-track#648]
### Current Behavior:

Dependency Track v3.7.0 runs fine for several days and then users are unable to login. The UI displays an error dialog "An unexpected error occurred...",

The dependency-track log displays no error, and all seems to still be OK with tasks such as MetricsUpdateTask, etc.  Automated BOM uploads from Jenkins (which use a token) are still fine.

Once the first login fail has occurred, The `server.log` starts to be spammed with `JDODataStoreException` that is obviously related to LDAP (see below).

The only solution is to restart the server.  This always works...  but I have now had to do it about 15 times.

### Steps to Reproduce:
For me, reproduction just involves leaving the server running long enough.

### Expected Behavior:
Stability! =)

### Environment:
  
  - Dependency-Track Version: 3.7.0  
  - Distribution: [ Executable WAR ]
  - BOM Format & Version: CycloneDX 1.1
  - Database Server: [ PostgreSQL ]
  - Browser: Latests Firefox or Chrome

Server authentication is set up using Microsoft Active Directory.

### Additional Details:
The same defect has been seen previously with DT v3.4.1 using MySQL.  See #422 

From `server.log`

```
2020-04-01 07:54:13,844 [] WARN [org.eclipse.jetty.server.HttpChannel] /api/v1/user/login
javax.servlet.ServletException: javax.jdo.JDODataStoreException: Iteration request failed : SELECT 'alpine.model.Team' AS "DN_TYPE","A1"."ID",,"A1"."NAME""A1"."NAME" AS "NUCORDER0"nullnullnull AS "NUCORDER0","A0"."MANAGEDUSER_ID" FROM "A0"."MANAGEDUSER_ID" FROM "MANAGEDUSERS_TEAMS" "A0" INNER JOIN "TEAM" "A1" ON "A0"."TEAM_ID" = "A1"."ID" WHERE EXISTS (SELECT 'alpine.model.ManagedUser''alpine.model.ManagedUser' AS "DN_TYPE",,"A0_SUB"."ID""A0_SUB"."ID" AS "DN_APPID" AS "DN_APPID"nullnullnull FROM "MANAGEDUSER" "A0_SUB" WHERE "A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID""A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID") ORDER BY EXISTS (SELECT 'alpine.model.ManagedUser''alpine.model.ManagedUser' AS "DN_TYPE",,"A0_SUB"."ID""A0_SUB"."ID" AS "DN_APPID" AS "DN_APPID"nullnullnull FROM "MANAGEDUSER" "A0_SUB" WHERE "A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID""A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID")"NUCORDER0" ORDER BY "NUCORDER0"
NestedThrowables:
org.postgresql.util.PSQLException: No value specified for parameter 2.
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:408)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:365)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:318)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsyncServlet.service(ServletHolder.java:1400)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:760)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
	at alpine.filters.ContentSecurityPolicyFilter.doFilter(ContentSecurityPolicyFilter.java:224)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at alpine.filters.ClickjackingFilter.doFilter(ClickjackingFilter.java:92)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at alpine.filters.WhitelistUrlFilter.doFilter(WhitelistUrlFilter.java:115)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1596)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:590)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1607)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1297)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1577)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1212)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:270)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:388)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.jdo.JDODataStoreException: Iteration request failed : SELECT 'alpine.model.Team' AS "DN_TYPE","A1"."ID",,"A1"."NAME""A1"."NAME" AS "NUCORDER0"nullnullnull AS "NUCORDER0","A0"."MANAGEDUSER_ID" FROM "A0"."MANAGEDUSER_ID" FROM "MANAGEDUSERS_TEAMS" "A0" INNER JOIN "TEAM" "A1" ON "A0"."TEAM_ID" = "A1"."ID" WHERE EXISTS (SELECT 'alpine.model.ManagedUser''alpine.model.ManagedUser' AS "DN_TYPE",,"A0_SUB"."ID""A0_SUB"."ID" AS "DN_APPID" AS "DN_APPID"nullnullnull FROM "MANAGEDUSER" "A0_SUB" WHERE "A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID""A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID") ORDER BY EXISTS (SELECT 'alpine.model.ManagedUser''alpine.model.ManagedUser' AS "DN_TYPE",,"A0_SUB"."ID""A0_SUB"."ID" AS "DN_APPID" AS "DN_APPID"nullnullnull FROM "MANAGEDUSER" "A0_SUB" WHERE "A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID""A0_SUB"."USERNAME" = ? AND "A0"."MANAGEDUSER_ID" = "A0_SUB"."ID")"NUCORDER0" ORDER BY "NUCORDER0"
	at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:542)
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:456)
	at org.datanucleus.api.jdo.JDOQuery.execute(JDOQuery.java:276)
	at alpine.persistence.AlpineQueryManager.getManagedUser(AlpineQueryManager.java:303)
	at alpine.auth.ManagedUserAuthenticationService.authenticate(ManagedUserAuthenticationService.java:72)
	at alpine.auth.Authenticator.authenticate(Authenticator.java:65)
	at org.dependencytrack.resources.v1.UserResource.validateCredentials(UserResource.java:91)
	at sun.reflect.GeneratedMethodAccessor340.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:392)
	... 44 common frames omitted
Caused by: org.postgresql.util.PSQLException: No value specified for parameter 2.
	at org.postgresql.core.v3.SimpleParameterList.checkAllParametersSet(SimpleParameterList.java:257)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:292)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:143)
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:106)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java)
	at org.datanucleus.store.rdbms.ParamLoggingPreparedStatement.executeQuery(ParamLoggingPreparedStatement.java:375)
	at org.datanucleus.store.rdbms.SQLController.executeStatementQuery(SQLController.java:545)
	at org.datanucleus.store.rdbms.query.JDOQLQuery.performExecute(JDOQLQuery.java:703)
	at org.datanucleus.store.query.Query.executeQuery(Query.java:1967)
	at org.datanucleus.store.query.Query.executeWithArray(Query.java:1856)
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:433)
	... 70 common frames omitted
```

**Comment by stevespringett** [comment]
No idea what's causing this. You can try setting the server to debug logging.

DT 3.8.0 and previous will attempt user login using managed users first, then ldap. So even when a user is in a directory and not local, DT it will attempt to login that user as a managed user first, then fall back to LDAP if configured. This is to allow the admin to login locally in the event that LDAP fails or is configured incorrectly.

So, this issue is likely unrelated to LDAP and more closely related to Postgres and the DT persistence functionality.

**Comment by stevespringett** [comment]
Is this still an issue?

**Comment by msymons** [comment]
It was still an occasional issue with DT 3.8.0.   We set up a new 4.x server in test just under 4 weeks ago and switched to using it as our production DT last week when 4.3.1 was released (with its' PostgreSQL fixes).   Now using 4.3.3.

I will keep an eye on things and let you know how things look.  This won't be immediate...  but the longer span of time that elapses without a problem, the more it will look like it's addressed.
